go/vt/sqlparser/ast.go

```diff
@@ -2561,6 +2640,61 @@ type (
 	JSONUnquoteExpr struct {
 		JSONValue Expr
 	}
+
+	// ArgumentLessWindowExpr stands for the following window_functions: CUME_DIST, DENSE_RANK, PERCENT_RANK, RANK, ROW_NUMBER
+	// These functions do not take any argument.
+	ArgumentLessWindowExpr struct {
+		Type       ArgumentLessWindowExprType
+		OverClause *OverClause
+	}
+
+	// ArgumentLessWindowExprType is an enum to get types of ArgumentLessWindowExpr.
+	ArgumentLessWindowExprType int8
+
+	// FirstOrLastValueExpr stands for the following window_functions: FIRST_VALUE, LAST_VALUE
+	FirstOrLastValueExpr struct {
+		Type                FirstOrLastValueExprType
+		Expr                Expr
+		NullTreatmentClause *NullTreatmentClause
+		OverClause          *OverClause
+	}
+
+	// FirstOrLastValueExprType is an enum to get types of FirstOrLastValueExpr.
+	FirstOrLastValueExprType int8
+
+	// NtileExpr stands for the NTILE()
+	NtileExpr struct {
+		IntValue   *int
+		VarValue   ColIdent
```

Comment:
Maybe it's just me, but I think I'd prefer `VarValue` to be a ref, and if both `VarValue` and `IntValue` are nil, we know we are dealing with a null.

The problem with `IsNull` here (and admittedly it's a minor one) is that you can construct a struct that has "inconsistent state". if `IsNull` is true _and_ `VarValue` has a non-empty value, how do you know what you have?

Reply:
How the values are passed to the struct in the parser is decided by us. So will it affect us much?

Reply:
well... now I would have to trust the people that wrote that parser, and you know they can't be trusted... ;)

like I wrote - it's not super-important. I just prefer to design structs so that they can't contain invalid data, unless there's a strong reason to do something different.

Reply:
We don't even need to make VarValue a ref, there is a IsEmpty() function on ColIdent

Reply:
I have opted for a different route. I stored a single value in the struct of type Expr instead of having 3 there of which only 1 should be specified. I think this is cleaner and easier to use down the line.

Reply:
One refactor which has been deferred for now, and probably out of scope of this PR is that, we are storing user-defined variables, in columnName struct which isn't super correct, but we can fix it later.
For example, `select @@x` will have @@x stored inside a colident inside a column name